File: iOS/Sources/Data/Repositories/InMemoryPocketRepository.swift
import Foundation

class InMemoryPocketRepository: PocketRepositoryProtocol {
    private var pockets: [Pocket] = []
    private var nextId: Int64 = 1

    func fetchAll() async throws -> [Pocket] {
        return pockets.sorted { $0.name < $1.name }
    }

    func fetch(id: Int64) async throws -> Pocket {
        guard let pocket = pockets.first(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        return pocket
    }

    func add(_ pocket: Pocket) async throws -> Pocket {
        let newPocket = Pocket(
            id: nextId,
            name: pocket.name,
            description: pocket.description,
            balance: pocket.balance,
            createdAt: pocket.createdAt,
            updatedAt: pocket.updatedAt
        )
        nextId += 1
        pockets.append(newPocket)
        return newPocket
    }

    func update(_ pocket: Pocket) async throws {
        guard let index = pockets.firstIndex(where: { $0.id == pocket.id }) else {
            throw DomainError.notFound
        }
        var updated = pocket
        updated.updatedAt = Date()
        pockets[index] = updated
    }

    func delete(id: Int64) async throws {
        guard pockets.contains(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        pockets.removeAll { $0.id == id }
    }

    func updateBalance(id: Int64, amount: Double) async throws {
        guard let index = pockets.firstIndex(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        pockets[index].balance += amount
        pockets[index].updatedAt = Date()
    }
}

File: android/app/src/main/java/com/budgetmanager/data/repository/InMemoryPocketRepository.kt
package com.budgetmanager.data.repository

import com.budgetmanager.domain.entity.DomainError
import com.budgetmanager.domain.entity.Pocket
import com.budgetmanager.domain.repository.PocketRepository
import java.time.LocalDateTime

class InMemoryPocketRepository : PocketRepository {
    private val pockets = mutableListOf<Pocket>()
    private var nextId: Long = 1

    override suspend fun getAll(): List<Pocket> {
        return pockets.sortedBy { it.name }
    }

    override suspend fun getById(id: Long): Pocket {
        return pockets.find { it.id == id } ?: throw DomainError.NotFound
    }

    override suspend fun add(pocket: Pocket): Pocket {
        val newPocket = pocket.copy(id = nextId++)
        pockets.add(newPocket)
        return newPocket
    }

    override suspend fun update(pocket: Pocket) {
        val index = pockets.indexOfFirst { it.id == pocket.id }
        if (index == -1) throw DomainError.NotFound

        pockets[index] = pocket.copy(updatedAt = LocalDateTime.now())
    }

    override suspend fun delete(id: Long) {
        if (!pockets.removeIf { it.id == id }) {
            throw DomainError.NotFound
        }
    }

    override suspend fun updateBalance(id: Long, amount: Double) {
        val index = pockets.indexOfFirst { it.id == id }
        if (index == -1) throw DomainError.NotFound

        val pocket = pockets[index]
        pockets[index] = pocket.copy(
            balance = pocket.balance + amount,
            updatedAt = LocalDateTime.now()
        )
    }
}
